core/src/main/kotlin/McuManager.kt

```diff
@@ -0,0 +1,142 @@
+package com.juul.mcumgr
+
+import com.juul.mcumgr.message.CoreReadRequest
+import com.juul.mcumgr.message.CoreReadResponse
+import com.juul.mcumgr.message.EchoRequest
+import com.juul.mcumgr.message.EchoResponse
+import com.juul.mcumgr.message.FileReadRequest
+import com.juul.mcumgr.message.FileReadResponse
+import com.juul.mcumgr.message.FileWriteRequest
+import com.juul.mcumgr.message.FileWriteResponse
+import com.juul.mcumgr.message.ImageWriteRequest
+import com.juul.mcumgr.message.ImageWriteResponse
+import com.juul.mcumgr.message.Request
+import com.juul.mcumgr.message.Response
+import com.juul.mcumgr.transfer.CoreDownloader
+import com.juul.mcumgr.transfer.DownloadProgress
+import com.juul.mcumgr.transfer.Downloader
+import com.juul.mcumgr.transfer.FileDownloader
+import com.juul.mcumgr.transfer.FileUploader
+import com.juul.mcumgr.transfer.ImageUploader
+import com.juul.mcumgr.transfer.UploadProgress
+import com.juul.mcumgr.transfer.Uploader
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+
+class McuManager(val transport: Transport) {
+
+    // System
+
+    suspend fun echo(echo: String): McuMgrResult<EchoResponse> =
+        send(EchoRequest(echo))
+
+    // Image
+
+    suspend fun imageWrite(
+        data: ByteArray,
+        offset: Int,
+        length: Int? = null,
+        hash: ByteArray? = null
+    ): McuMgrResult<ImageWriteResponse> =
+        send(ImageWriteRequest(data, offset, length, hash))
+
+    suspend fun coreRead(offset: Int): McuMgrResult<CoreReadResponse> =
+        send(CoreReadRequest(offset))
+
+    // Files
+
+    suspend fun fileWrite(
+        fileName: String,
+        data: ByteArray,
+        offset: Int,
+        length: Int? = null
+    ): McuMgrResult<FileWriteResponse> {
+        return send(FileWriteRequest(fileName, data, offset, length))
+    }
+
+    suspend fun fileRead(fileName: String, offset: Int): McuMgrResult<FileReadResponse> =
+        send(FileReadRequest(fileName, offset))
+
+    // Upload
+
+    suspend fun uploadImage(
+        data: ByteArray,
+        windowCapacity: Int = 1,
+        progressHandler: ((UploadProgress) -> Unit)? = null
+    ): McuMgrResult<Unit> {
+        val uploader = ImageUploader(data, this, windowCapacity)
+        return catchResult {
+            upload(uploader, progressHandler)
+        }
+    }
+
+    suspend fun uploadFile(
+        data: ByteArray,
+        fileName: String,
+        windowCapacity: Int = 1,
+        progressHandler: ((UploadProgress) -> Unit)? = null
+    ): McuMgrResult<Unit> {
+        val uploader = FileUploader(data, fileName, this, windowCapacity)
+        return catchResult {
+            upload(uploader, progressHandler)
+        }
+    }
+
+    @Throws
+    suspend fun upload(
+        uploader: Uploader,
+        progressHandler: ((UploadProgress) -> Unit)? = null
```

Comment:
This feels like Java callback style when you have Coroutines to provide data streams with all its fancy operators.

In fact, it looks like you're wrapping a nice Coroutines API in this callback style function.

This function feels like a good candidate for a separate module where you bridge the Coroutine and Java world. Doesn't feel like it belongs in this module.

Instead, you could have:

```kotlin
class McuManager {
    // ...

    fun fileUploader(
        data: ByteArray,
        fileName: String,
        windowCapacity: Int = 1
    ) = FileUploader(data, fileName, this, windowCapacity)

    // ...
}
```

Users of library can easily call and use it as they see fit:

```kotlin
val mcumgr: McuManager

// When progress isn't needed.
mcumgr.fileUploader(...).upload() // Suspends until complete.
```

If progress is desired then one possible usage could look like:

```kotlin
val mcumgr: McuManager
val uploader = mcumgr.fileUploader(...)
uploader
    .progress
    .transformWhile { progress ->
        emit(progress)
        progress != 100
    }
    .onEach { ... }
    .launchIn(scope)
uploader.upload() // Suspends until complete.
```